### PR TITLE
[AIRFLOW-5377] Add Google Campaign Manager operators

### PR DESCRIPTION
**Pull request by nuclearpinguin**
- [x] CampaignManagerCreateReportOperator - Creates a new report in Campaign Manager
- [x] CampaignManagerRunReportOperator - Runs a Campaign Manager report
- [x] CampaignManagerReportSensor - Waits until a Campaign Manager report is ready for download
- [x] CampaignManagerDownloadReportOperator - Downloads a completed Campaign Manager report file
- [x] CampaignManagerDeleteReportOperator - Deletes an existing Campaign Manager report

### CODE REVIEW

**Comment by TobKed** [comment]
LGTM

File: airflow/gcp/example_dags/example_campaign_manager.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    "example_campaign_manager",
+    default_args=default_args,
+    schedule_interval=None,  # Override to match your needs
+) as dag:
+    # [START howto_campaign_manager_insert_report_operator]
+    create_report = GoogleCampaignManagerInsertReportOperator(
+        profile_id=PROFILE_ID, report=REPORT, task_id="create_report"
+    )
+    report_id = "{{ task_instance.xcom_pull('create_report')['id'] }}"
+    # [END howto_campaign_manager_insert_report_operator]
```
Comment (mik-laj) [inline airflow/gcp/example_dags/example_campaign_manager.py]: Is the end marker in the right place?

File: airflow/gcp/hooks/campaign_manager.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+            .reports()
+            .list(
+                profileId=profile_id,
+                maxResults=max_results,
+                pageToken=page_token,
+                scope=scope,
+                sortField=sort_field,
+                sortOrder=sort_order,
+            )
+            .execute(num_retries=self.num_retries)
```
Comment (mik-laj) [inline airflow/gcp/hooks/campaign_manager.py]: Is pagination missing here?

File: airflow/gcp/hooks/campaign_manager.py [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+        :type file_id: str
+        :return: googleapiclient.http.HttpRequest
+        """
+        response = (
+            self.get_conn()  # pylint: disable=no-member
+            .reports()
+            .files()
+            .get_media(fileId=file_id, profileId=profile_id, reportId=report_id)
+        )
+        return response
```
Comment (mik-laj) [inline airflow/gcp/hooks/campaign_manager.py]: Can you add the wait for operation method? This is not needed, but it is useful from the point of view of the whole hook.

File: airflow/gcp/operators/campaign_manager.py [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+    :param api_version: The version of the api that will be requested for example 'v3'.
+    :type api_version: str
+    :param gcp_conn_id: The connection ID to use when fetching connection info.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any. For this to work, the service accountmaking the
+        request must have  domain-wide delegation enabled.
+    :type delegate_to: str
+    """
+
+    template_fields = ("profile_id", "report_id")
```
Comment (mik-laj) [inline airflow/gcp/operators/campaign_manager.py]: Can you add more template_fields?

File: airflow/gcp/operators/campaign_manager.py [hunk 5]
```diff
@@ -0,0 +1,312 @@
...
+        self.gcp_conn_id = gcp_conn_id
+        self.delegate_to = delegate_to
+
+    @staticmethod
+    def _set_report_name(name: str) -> str:
+        return name if name.endswith(".csv") else name + ".csv"
+
+    @staticmethod
+    def _set_bucket_name(name: str) -> str:
+        bucket = name if not name.startswith("gs://") else name[5:]
```
Comment (mik-laj) [inline airflow/gcp/operators/campaign_manager.py]: I think, we should make _parse_gcs_url publicly and use it here

File: airflow/gcp/operators/campaign_manager.py [hunk 6]
```diff
@@ -0,0 +1,312 @@
...
+    :type api_version: str
+    :param gcp_conn_id: The connection ID to use when fetching connection info.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any. For this to work, the service accountmaking the
+        request must have  domain-wide delegation enabled.
+    :type delegate_to: str
+    """
+
+    template_fields = ("profile_id", "report")
+    template_ext = (".json",)
```
Comment (mik-laj) [inline airflow/gcp/operators/campaign_manager.py]: Do we have an example that uses this?

File: docs/howto/operator/gcp/campaign_manager.rst [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+.. exampleinclude:: ../../../../airflow/gcp/example_dags/example_campaign_manager.py
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_campaign_manager_run_report_operator]
+    :end-before: [END howto_campaign_manager_run_report_operator]
+
+You can use :ref:`Jinja templating <jinja-templating>` with
+:template-fields:`airflow.gcp.operators.campaign_manager.GoogleCampaignManagerRunReportOperator`
+parameters which allows you to dynamically determine values.
+The result is saved to :ref:`XCom <concepts:xcom>`, which allows it to be used by other operators.
```
Comment (mik-laj) [inline docs/howto/operator/gcp/campaign_manager.rst]: Can you add reference also?

File: tests/gcp/sensors/test_campaign_manager.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+# under the License.
+from unittest import TestCase, mock
+
+from airflow.gcp.sensors.campaign_manager import GoogleCampaignManagerReportSensor
+
+API_VERSION = "api_version"
+GCP_CONN_ID = "google_cloud_default"
+
+
+class TestCampaignManagerDeleteReportOperator(TestCase):
```
Comment (mik-laj) [inline tests/gcp/sensors/test_campaign_manager.py]: Can you check this name?

File: airflow/gcp/example_dags/example_campaign_manager.py (hunk 1 above)
Comment (nuclearpinguin) [reply airflow/gcp/example_dags/example_campaign_manager.py]: Yes, I wanted to include how to get information from xcom.

File: airflow/gcp/operators/campaign_manager.py (hunk 5 above)
Comment (nuclearpinguin) [reply airflow/gcp/operators/campaign_manager.py]: I agree, should it go to utils or to some hook as a static method?

File: docs/howto/operator/gcp/campaign_manager.rst (hunk 7 above)
Comment (nuclearpinguin) [reply docs/howto/operator/gcp/campaign_manager.rst]: Reference to what?

File: airflow/gcp/hooks/campaign_manager.py [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+        response = (
+            self.get_conn()  # pylint: disable=no-member
+            .reports()
+            .files()
+            .get(fileId=file_id, profileId=profile_id, reportId=report_id)
+            .execute(num_retries=self.num_retries)
+        )
+        return response
+
+    def get_report_file(
```
Comment (TobKed) [inline airflow/gcp/hooks/campaign_manager.py]: This method is not covered in unit tests. Could you add tests?

File: airflow/gcp/sensors/campaign_manager.py [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+    :type profile_id: str
+    :param report_id: The ID of the report.
+    :type report_id: str
+    :param file_id: The ID of the report file.
+    :type file_id: str
+    :param api_version: The version of the api that will be requested for example 'v3'.
+    :type api_version: str
+    :param gcp_conn_id: The connection ID to use when fetching connection info.
+    :type gcp_conn_id: str
+    :param delegate_to: The account to impersonate, if any. For this to work, the service accountmaking the
```
[the remark this answers is not in the document]
Comment (nuclearpinguin) [reply airflow/gcp/sensors/campaign_manager.py]: I am not sure if this is a name or a mistake 🤷🏼‍♂️ Was taken from API description.

File: airflow/gcp/hooks/campaign_manager.py (hunk 2 above)
Comment (nuclearpinguin) [reply airflow/gcp/hooks/campaign_manager.py]: Added

File: airflow/gcp/operators/campaign_manager.py (hunk 4 above)
Comment (nuclearpinguin) [reply airflow/gcp/operators/campaign_manager.py]: Done

File: airflow/gcp/hooks/campaign_manager.py [hunk 11]
```diff
@@ -0,0 +1,229 @@
...
+    Hook for Google Campaign Manager.
+    """
+
+    _conn = None  # type: Optional[Resource]
+
+    def __init__(
+        self,
+        api_version: str = "v3.2",
+        gcp_conn_id: str = "google_cloud_default",
+        delegate_to: str = None,
```
[the remark this answers is not in the document]
Comment (nuclearpinguin) [reply airflow/gcp/hooks/campaign_manager.py]: Good catch!

File: airflow/gcp/operators/campaign_manager.py (hunk 6 above)
Comment (nuclearpinguin) [reply airflow/gcp/operators/campaign_manager.py]: Good catch, added ;)

File: airflow/gcp/hooks/campaign_manager.py (hunk 2 above)
Comment (nuclearpinguin) [reply airflow/gcp/hooks/campaign_manager.py]: Added

**Pull request closed without merging**